[AH-8] erd 초안 작성

Yjooon opened the pull request:
## 개요
- 초기 JPA 엔티티를 작성했습니다.

## 변경 사항
- 주문/그룹
- [x] GroupOrder, GroupOrderItem, GroupOrderMember 작성

- 결제/지갑
- [x] PaymentIntent, PayMoneyWallet, WalletHold, WalletLedger 작성

- 매장/메뉴
- [x] Restaurant(초안), MenuItem(초안) 작성

- 사용자/픽업
- [x] SolUser(초안), PickupZone(초안) 작성

### 추가 변경 사항
- [x] BaseEntity 작성


## 설명
초기 JPA 엔티티를 작성했습니다.
공통된 createdAt, updatedAt, 부분은 BaseTimeEntity에 작성했습니다.
모든 entity의 동등성을 pk로 구분하게 작성했습니다. 
identity / sequence 구분하여 상속받을 수 있게 BaseEntity 작성했습니다.
코드 리뷰의 수정 사항 반영했습니다.

## 제외
order 부분 entity의 동시성 관리, 실패 사유 저장 필드 추가 작업은 #7 에서 진행합니다.

## 관련 이슈
- resolves #4

Yjooon committed:
feat: 지갑, 결제 ERD 작성

Yjooon committed:
feat: user ERD 초안 작성

Yjooon committed:
feat: 그룹 주문 및 식당, 메뉴, 픽업존 ERD 작성

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/campus/model/entity/PickupZone.java:
```diff
@@ -0,0 +1,36 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+@Table(name = "pickup_zone",
+        uniqueConstraints = @UniqueConstraint(name = "uk_pickup_zone_name", columnNames = {"campus_code", "label"}),
```
campus_code랑 label 조합이 유일해야한다고 제약 조건을 거신것 같은데, label 이라는 필드가 없는 것 같습니다. 제가 이 코드를 잘 이해한것이 맞을까요??

Yjooon replied:
초기 설계에 지도에 마커 표시를 label로 관리하려고 분리했는데, 실제 사용처가 확실하지 않아서 제거했는데 깜빡하고 안지웠습니다. 수정하겠습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/order/model/entity/GroupOrder.java:
```diff
@@ -0,0 +1,81 @@
+    @Column(nullable = false, length = 16)
+    private Status status = Status.FORMING;
+
+    @Column(name = "created_at",
+            columnDefinition = "timestamptz NOT NULL DEFAULT now()",
+            insertable = false, updatable = false)
+    private OffsetDateTime createdAt;
```
createdAt 과 같은 필드까지 꼼꼼하게 설계해주신거 좋은 것 같습니다.
create,update 시점 같은 필드는 모든 엔티티에 거의 공통적으로 들어갈 것 같은데 상속을 고려해보는 것은 어떨까요

kwonssshyeon replied:
그리고 `timestamptz` 은 postgre에 종속적인 문법인것 같네유

Yjooon replied:
상속 코드로 수정하겠습니다.

다른 DB로의 이식성은 고려 안하고 작성했어서,,, 수정하겠습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/campus/model/entity/PickupZone.java:
```diff
@@ -0,0 +1,36 @@
+    private String name;
+
+    // 지도 마커 표시용 좌표
+    @Column(name = "lat", nullable = false)
+    private double lat;
+
+    @Column(name = "lng", nullable = false)
+    private double lng;
```
이부분은 wrapper 타입으로 수정하는 건 어떨까요?
사소하긴 하지만, 빈 값인 경우 double이면 초기값으로 0.0이 들어가서 null이라는 의미와는 다른 결과를 줄 수 있을 것 같습니다.
대신 null과 관련한 처리를 직접 처리해야한다는 번거로움은 있을 것 같아요. 객체 생성 시에 반드시 채워져야 하는 필드이고 null인 경우가 없는게 확실하다면 지금처럼 원시타입으로 작성해도 좋을 것 같아요!

Yjooon replied:
wrapper 타입이 더 적합할 것 같네요. 수정하겠습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/order/model/entity/GroupOrderItem.java:
```diff
@@ -0,0 +1,48 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@ToString(exclude = {"member","menuItem"})
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
```
이 엔티티에서만 equals랑 hashcode 비교 필드를 직접 정한 이유가 궁금합니다 !

kwonssshyeon replied:
제가 이 어노테이션을 처음봤는데 제가 찾아본 바로는, 새로 생성된 객체에서 id가 null인 경우 동일 객체로 잘못 판단할 수 있어, null 체크 후 id를 비교하도록 equals/hashCode를 직접 오버라이딩하는 방식이 안전하다고 합니다.
제가 대략적으로 확인한 내용이라, 놓친 부분이 있을 수도 있습니다.
한번 확인해주세용

Yjooon replied:
GroupOrderItem, GroupOrderMember는 각각 다른 엔티티에서 List에 포함되고 있어서, equals를 id 기준으로 정의해주지 않으면 List의 remove 와 같은 작업에서 프록시 객체를 다룰 때 문제가 발생할 수 있을 것 같아서 명시해두었습니다. 

id 가 null인 경우 동일 객체로 잘못 판단할 수 있는 부분은 확인했습니다. 수정하겠습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/user/model/entity/SolUser.java:
```diff
@@ -0,0 +1,25 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "student_id")
+    private Long studentId;
```
이거 학번인가요? 학번이라면 String 으로 하는게 좋을 것 같아요 !
제가 저번에 학번으로 뭐 구현할때 교환학생 학번이 `B12345` 이런 형태여서 대참사가 났던 경험이 있습니다 ..

Yjooon replied:
제 학교에선 전부 숫자였어서 알파벳 섞인 학번도 있는 줄 몰랐네요 수정하겠습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/wallet/model/entity/PayMoneyWallet.java:
```diff
@@ -0,0 +1,42 @@
+    @Column(name = "amount", nullable = false)
+    private long amount;
+
+    @Version
+    @Column(nullable = false)
+    private long version;
```
동시성 고려해주신점 좋은 것 같습니다. !
이 부분은 금액과 관련된 부분인 만큼 비관락을 사용하는 것은 어떨까요?
예를 들어, 돈을 출금,송금한 후에 버전체크를 해서 결제에 실패했습니다. -> 이렇게 하는 것 보다는
출금/송금하기전에, '송금 작업중' -> 이런식으로 선점되었음을 알리는 것이 이 엔티티에 대해서는 적합할 것 같다는 생각이 듭니다.

Yjooon replied:
말씀주신 방법이 더 적합한 것 같네요. 수정하겠습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/wallet/model/entity/WalletLedger.java:
```diff
@@ -0,0 +1,77 @@
+        },
+        uniqueConstraints = @UniqueConstraint(name = "uk_ledger_idempotency", columnNames = "idempotency_key")
+)
+@Immutable
```
변경 불가능하게 막은거 좋은것 같습니다.

kwonssshyeon commented on src/main/java/com/alddeul/solsolhanhankki/wallet/model/entity/WalletLedger.java:
```diff
@@ -0,0 +1,77 @@
+        uniqueConstraints = @UniqueConstraint(name = "uk_ledger_idempotency", columnNames = "idempotency_key")
+)
+@Immutable
+@Check(constraints = "amount > 0")
```
amount는 0 이 될 수는 없는 건가요 ?
필드의 역할을 잘 파악하지 못해 여쭤봅니다.

Yjooon replied:
지갑 액수의 변동 내역을 관리하는 엔티티라서, amount가 0이면 변동 사항이 없다는 뜻이니 변동 기록이 존재할 수 없다고 생각되어 amount > 0 제약을 걸었습니다.

Yjooon committed:
Merge branch 'develop' into feat/erd-AH-8

kwonssshyeon reviewed (commented):
엔티티 설계를 꼼꼼하게 하신게 엄청 잘 느껴지는 것 같습니다. 수고하셨습니다
- 그리고 공통적으로 id를 IDENTIFY를 사용하셨는데, SEQUENCE 사용도 고려해보는 것도 좋을 것 같습니다 ! 특히, db round trip이나 배치업데이트가 있을만한곳에는 도입하는 것도 좋을것 같습니다. 관리해야할 테이블이 많이 늘어난다는 단점이 있을 거 같긴하네용...
- version 동시성 제어, wrapper/primitive 사용은 리뷰하지 않은 다른 필드에 대해서 고려해보는 것도 좋을 것 같습니다.
- 인덱스나 연관관계 매핑까지는 꼼꼼히 확인하지 못했습니다 ㅠㅠ
- 구조를 이해하지 못한 부분이 몇 군데 있는데, 내일 여쭤보겠습니다 !

진짜 수고많으셨어용... 바톤 터치하겠습니다 ....!


+) 실패 상태일때는 왜 실패인지 기록할 수 있는 필드도 있으면 좋을 것 같슴다

Yjooon committed:
feat: BaseEntity 추가 및 JPA Auditing 적용

createdAt, updatedAt 필드를 BaseTimeEntity에 통합했습니다.
identity 적용 Entity는 BaseIdentityEntity,
sequence 적용 Entity는 BaseSeqEntity 로 분리하여 상속받을 수 있게 구현했습니다.

Yjooon committed:
refactor: 각 entity에 BaseEntity 적용

Yjooon committed:
feat: label 제거, 좌표 값 wrapper 타입으로 변경

Yjooon committed:
feat: 개설자 삭제

Yjooon committed:
feat: 학번 Long -> String으로 변경
